pipeline/photo_cal.py

```diff
@@ -28,12 +70,185 @@ class PhotCalibrator:
+           The uncertainty on the zeropoint
+
         """
-        Extract sources and use their magnitudes to calculate the photometric solution.
+
+        # For GaiaDR3, catexp.data has fields:
+        #   X_WORLD
+        #   Y_WORLD
+        #   MAG_G
+        #   MAG_BP
+        #   MAG_RP
+        #   MAGERR*
+
+        # Extract from the catalog the range of Gaia colors that we want
+        # to use (motivated by looking at Gaia H-R diagrams).
+
+        gaiaminbp_rp = 0.5
+        gaiamaxbp_rp = 3.0
+        col = catexp.data[ 'MAG_BP' ] - catexp.data[ 'MAG_RP' ]
+        catdata = catexp.data[ ( col >= gaiaminbp_rp ) & ( col <= gaiamaxbp_rp ) ]
+        _logger.debug( f"{len(catdata)} Gaia stars with {gaiaminbp_rp}<BP-RP<{gaiamaxbp_rp}" )
+
+        # Pull out the source information. Use the aperture that the
+        # source list has designated as the "infinite" aperture, so that
+        # there's no need to add an aperture correction into the
+        # calculated zeropoint.
+
+        sourceflux, sourcefluxerr = sources.apfluxadu( apnum=sources.inf_aper_num )
+        skycoords = wcs.wcs.pixel_to_world( sources.x, sources.y )
+        sourcera = skycoords.ra.deg
+        sourcedec = skycoords.dec.deg
+
+        # Only use stars that are "good" and that have flux > 3*fluxerr
+
+        wgood = ( sources.good & sources.is_star
+                  & ( ~np.isnan( sourceflux) ) & ( ~np.isnan( sourcefluxerr ) )
+                  & ( sourceflux > 3.*sourcefluxerr ) )
+        sourcera = sourcera[wgood]
+        sourcedec = sourcedec[wgood]
+        sourceflux = sourceflux[wgood]
+        sourcefluxerr = sourcefluxerr[wgood]
+        _logger.debug( f"{len(sourcera)} of {sources.num_sources} image sources chosen; "
+                       f"{(sources.good & sources.is_star).sum()} are good stars, of those "
+                       f"{len(sourcera)} have flux in 'infinite' aperture >3σ" )
+
+        # Match catalog excerpt RA/Dec to source RA/Dec
+
+        catdex = []
+        sourcedex = []
+        # TODO : get rid of for loop and use array function
+        for i in range( len(catdata) ):
+            # TODO : the 360 to 0° issue that we have to deal with in a huge number of places
+            # Astropy.Longitude could come to our rescue?
+            dra = sourcera - catdata[i]['X_WORLD']
+            ddec = sourcedec - catdata[i]['Y_WORLD']
+            w = np.where( ( np.fabs( dra ) < match_size/3600./np.cos( sourcedec ) )
+                          & ( np.fabs( ddec ) < match_size/3600. ) )[0]
+            if len(w) == 0:
+                continue
+            else:
+                # If there are more than one, just pick the first one in the list
+                # that matches.
+                # TODO : think about whether we can do better.
+                catdex.append( i )
+                sourcedex.append( w[0] )
+
+        _logger.debug( f"Matched {len(catdex)} stars between Gaia and the image source list" )
+
```

Comment:
We can probably improve this using `argmin` and then a check if it is inside the match radius. 

In the interest of "if it's not broken don't fix it" I'll make an issue to come back and improve this part later.